core/src/main/java/org/elasticsearch/common/lucene/all/AllTermQuery.java

```diff
@@ -124,14 +125,21 @@ public boolean equals(Object obj) {
 
     @Override
     public Query rewrite(IndexReader reader) throws IOException {
+        boolean fieldExists = false;
         boolean hasPayloads = false;
         for (LeafReaderContext context : reader.leaves()) {
             final Terms terms = context.reader().terms(term.field());
-            if (terms.hasPayloads()) {
-                hasPayloads = true;
-                break;
+            if (terms != null) {
+                fieldExists = true;
+                if (terms.hasPayloads()) {
+                    hasPayloads = true;
+                    break;
+                }
             }
         }
+        if (fieldExists == false) {
+            return new MatchNoDocsQuery();
```

Comment:
Can you propagate the boost to this query? If this query is enclosed in a BooleanQuery, it could have an impact on the normalization factor.